docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,213 @@
+
+When the profile is applied, `tedge-agent` is free to group modules of the same kind that are listed sequentially,
+to optimize the execution of that operation.
+For example, if 3 software modules are defined sequentially, the agent could group them into an `updateList`
+so that they can be installed in one-go using the `update-list` API of software plugins.
+
+## Tedge agent handling device profile commands
+
+The `tedge-agent` handless `device_profile` commands using the workflow definition at `/etc/tedge/operations/device_profile.toml`.
+This workflow definition handles each module type using sub-operation workflows defined for that type.
+For example, the firmware module is installed by triggering a `firmware_update` sub-command
+which in turn uses the `firmware_update` workflow for that operation execution.
+Similarly software modules are installed with `software_update` subcommands and
+configuration updates are applied using `config_update` subcommands.
+These subcommands are triggered for each module defined in the profile definition in that order.
+
+### On success
+
+Once all the modules in the profile are successfully installed, the successful status is published
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile/1234' '{
+  "status": "successful",
+  ... // Other input fields
+}'
+```
+
+...and the current applied device profile information is updated by publishing the same to the capability topic as follows:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile' '{
+  "name": "prod-profile",
+  "version: "v2"
+}'
+```
+
+### On failure
+
+On failure, the `device_profile` operation is aborted at the module that caused the failure.
+The remaining modules in the profile are not installed and no attempt is made to rollback the already installed modules either.
+If the sub-operations support rollbacks at the sub-operation level, it is performed for the failed module.
+
+For example, if a profile includes a firmware, 2 software and 1 configuration updates in that sequence,
+if the failure happens during the second software update, no rollback is performed at the overall `device_profile` operation level,
+or even for that failed software update, unless the `software_update` workflow for that software type supports rollbacks.
+In that case the firmware and 1st software would remain installed, with the failed software update and last config update skipped.
+
+The `device_profile` operation itself is marked failed as follows:
+
+```sh te2mqtt formats=v1
+tedge mqtt pub -r 'te/device/main///cmd/device_profile/1234' '{
+  "status": "failed",
+  "reason": "Installation of software module: `jq` failed. Refer to operation logs for further details."
+  ... // Other input fields
+}'
+```
+
+The same profile can be applied again, by excluding all the modules(`"include": false`) before the failed `jq` module as follows:
```

Comment:
Does this have to be the *same* profile? I would expect the user to tune something, say changing some config settings or a software version.

Reply:
It doesn't **have to be** the same profile. What I meant is that the user doesn't always have to create a new profile to fix the issue, as the underlying issue can even be fixed externally (directly or via remote access plugin, may be), if the failure was limited to a small number of devices in the whole fleet. And then the same profile can be reapplied again. But, if he wishes to push the fix also via a device profile, then he can definitely do that, by having a newer version of the profile with the fix.

Obviously, if the failure was affecting the majority of the fleet, then the user would definitely want to push the fix also using a device profile.

Reply:
What is proposed by @reubenmiller in this [comment](https://github.com/thin-edge/thin-edge.io/pull/2920#discussion_r1648507189) seems simpler and more convenient for the users:

> We should let each firmware/software/configuration handler be responsible for the idempotency. The device profile should just execute the same workflow again, and each individual handler can work out if it should really do something or not.

Reply:
Resolved by pushing the idempotency requirement to the sub-operation and not mandating the user to skip the already applied operations.